frontend/setupChelonia.js

```diff
@@ -12,6 +12,43 @@ import manifests from './model/contracts/manifests.json'
 import { SETTING_CHELONIA_STATE, SETTING_CURRENT_USER } from './model/database.js'
 import { CHATROOM_USER_STOP_TYPING, CHATROOM_USER_TYPING, CHELONIA_STATE_MODIFIED, KV_EVENT, LOGGING_OUT, LOGIN_COMPLETE, LOGOUT, OFFLINE, ONLINE, RECONNECTING, RECONNECTION_FAILED, SERIOUS_ERROR } from './utils/events.js'
 
+const handleDeletedContract = async (contractID: string) => {
+  const { cheloniaState, contractState } = sbp('chelonia/contract/fullState', contractID)
+  if (!cheloniaState) return
+
+  await sbp('chelonia/contract/remove', contractID)
+
+  const type = cheloniaState.type?.replace(/^gi\.contracts\//, 'gi.actions/')
+  const handler = type && sbp('sbp/selectors/fn', `${type}/_ondeleted`)
+
+  const currentIdentityState = sbp('state/vuex/getters').currentIdentityState
+  // Delete redudant file deletion tokens. If a contract has been deleted, so
+  // have been the files it contained, and therefore we no longer need to
+  // hold on to their file deletion tokens.
+  if (currentIdentityState.fileDeleteTokens) {
+    // $FlowFixMe[incompatible-use]
+    const manifestCids = Object.entries(currentIdentityState.fileDeleteTokens).filter(([, { billableContractID }]) => {
+      return billableContractID === contractID
+    }).map(([cid]) => cid)
+    await sbp('gi.actions/identity/removeFiles', {
+      manifestCids,
+      option: {
+        shouldDeleteToken: true
+      }
+    }).catch(e => {
+      console.error('[handleDeletedContract] Error deleting saved tokens for deleted contract', contractID, e)
+    })
+  }
+
+  if (typeof handler === 'function') {
+    await handler(contractID, contractState).catch(e => {
+      console.error('Error handling deletion of contract', contractID)
+    })
```

Comment:
We need some clarification in a comment somewhere, perhaps here, or perhaps elsewhere, to explain why exactly this is safe to do like this...

The specific concern that I have is that when a group is deleted, it will result in these handlers being called on clients in an unpredictable order. In other words, perhaps the identity contract's `_ondeleted` will be called first, or last, or somewhere in between, to the `chatroom` and `group` `_ondeleted` calls... how do you know that no matter what order they're called in the same result will occur?

Reply:
That's a valid concern, but AFAICT the way it's written / used it's not really an issue, other than redundancy. For example, the `_ondeleted` on a chatroom will attempt to remove it from the parent group or identity contract. If the parent is what happened to also be deleted, this action will fail, but it's of little consequence as everything will be removed in the end.